internal/jsonutil/reader.go

```diff
@@ -0,0 +1,126 @@
+package jsonutil
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"iter"
+)
+
+// StreamArrayReader reads a JSON array from r and yields each element.
+func StreamArrayReader[T any](r io.Reader, keys []string) iter.Seq2[T, error] {
+	return func(yield func(T, error) bool) {
+		var v T
+		decoder := json.NewDecoder(r)
+
+		err := traverseKeys(decoder, keys)
+		if err != nil {
+			yield(v, err)
+			return
+		}
+
+		tk, err := decoder.Token()
+		if err != nil {
+			yield(v, fmt.Errorf("reading next token: %w", err))
+			return
+		}
+		if tk != json.Delim('[') {
+			yield(v, fmt.Errorf("expected `[`, got `%v`", tk))
+			return
+		}
+
+		for decoder.More() {
+			err := decoder.Decode(&v)
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			if !yield(v, err) {
+				return
+			}
+			if err != nil {
+				return
+			}
+		}
+
+		tk, err = decoder.Token()
+		if err != nil {
+			yield(v, fmt.Errorf("reading next token: %w", err))
+			return
+		}
+		if tk != json.Delim(']') {
+			yield(v, fmt.Errorf("expected `[`, got `%v`", tk))
+			return
+		}
+	}
+}
+
+func skipObject(decoder *json.Decoder) error {
+	t, err := decoder.Token()
+	if err != nil {
+		return err
+	}
+
+	_, ok := t.(json.Delim)
+	if !ok {
+		return nil
+	}
+
+	for decoder.More() {
+		err := skipObject(decoder)
```

Comment:
I'm having a hard time following this. Is this discarding all tokens between a matching pair of delimiters?

Do we need to worry about stack exhaustion here? Would it be better to increment a depth counter when we see a `[` or `{` token and decrement when we see a `]` or `}` token?

Reply:
Yes i.e. if you're inside an array you skip entities.

Reply:
And do we need to worry about stack depth when calling `skipObject()` recursively?

Reply:
this may be a concern for a general purpose library processing an unknown input, which is not the case for this utility. we may address that as a follow-up.